### PR TITLE
docs(tutorial): add a tutorial for the Flink backend

### PR DESCRIPTION
**Pull request by chloeh13q**
## Description of changes

Add a tutorial for the Flink backend.

### CODE REVIEW

File: docs/tutorials/open-source/apache-flink/1_single_feature.qmd [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+```
+
+The Kafka connector isn’t part of the binary distribution. In order to connect to a Kafka source/sink, we need to [download the JAR file](https://repo.maven.apache.org/maven2/org/apache/flink/flink-sql-connector-kafka/3.0.2-1.18/flink-sql-connector-kafka-3.0.2-1.18.jar) and manually add it into the classpath:
+
+```{python}
+# | code-fold: true
+!wget -N https://repo.maven.apache.org/maven2/org/apache/flink/flink-sql-connector-kafka/3.0.2-1.18/flink-sql-connector-kafka-3.0.2-1.18.jar
+```
+```{python}
+connection._exec_sql("ADD JAR './flink-sql-connector-kafka-3.0.2-1.18.jar'")
```
Comment (chloeh13q) [inline docs/tutorials/open-source/apache-flink/1_single_feature.qmd:67]: [Z] Use of a private method here, intended? The blog post will often be referred to as public documentation, i think we need more rigor here.

**Comment by ncclementi** [comment]
@chloeh13q for the[linting error](https://github.com/ibis-project/ibis/actions/runs/7637990693/job/20807965560?pr=8085) you are getting, you can run the pre-commit and that should fix them. https://ibis-project.org/contribute/03_style

**Comment by ncclementi** [comment]
The remaining CI failure is because `pyflink` is missing in the docs dependencies. We should add it to the group using poetry
```bash
poetry add --group docs pyflink
```

Then we will want to  run 
```bash 
poetry lock --no-upgrade
``` 

to update the lock file, then push all those changes up. I think we also might need to re-create the `requirements-dev.txt` but I'm not sure, maybe @gforsyth can chime in here?

**Comment by gforsyth** [comment]
If you have https://github.com/casey/just installed, you can run `just lock` to handle the relock and regeneration of the requirements dev file,  Otherwise, these are the commands to run:

```
poetry lock --no-update
poetry export --extras all --with dev --with test --with docs --without-hashes --no-ansi > requirements-dev.txt
```

**Comment by cpcloud** [comment]
This is unlikely to work, due to docs being built with nix. I suggest freezing the output for now until we can get this working properly.

File: docs/tutorials/open-source/apache-flink/0_setup.qmd [hunk 2]
```diff
@@ -0,0 +1,35 @@
+# Getting started
```
Comment (gforsyth) [inline docs/tutorials/open-source/apache-flink/0_setup.qmd:1]:
```suggestion
---
execute:
  freeze: auto
---

# Getting started
```

to freeze the output as it gets generated on your machine you can add this to the top of each of the `qmd` files

File: docs/tutorials/open-source/apache-flink/1_single_feature.qmd [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+
+
+### Prerequisites
+
+* Docker Compose: This tutorial uses Docker Compose to manage an Apache Kafka environment (including sample data generation) and a Flink cluster (for remote execution). You can [download and install Docker Compose from the official website](https://docs.docker.com/compose/install/).
+* JDK 11 release: Flink requires Java 11.
+* Python 3.9 or 3.10.
+* Follow [the instructions on the Ibis project page](https://ibis-project.org/install) to install the Flink backend for Ibis. For the tutorial below, we assume that you already have the Ibis package correctly installed in your environment.
+* Clone the [example repository](https://github.com/claypotai/ibis-flink-example).
```
Comment (chloeh13q) [inline docs/tutorials/open-source/apache-flink/1_single_feature.qmd:12]: We're hosting the example repo under our github account right now, but will most likely need to move this somewhere else (under ibis-project?), clean up some of the code, and figure out how best to structure multiple tutorials and blog posts.

File: docs/tutorials/open-source/apache-flink/0_setup.qmd [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+## Overview
+
+[Ibis](https://ibis-project.org/) is a Python library for writing and scaling data transformations. With close to 20 supported backends today, it allows you to write transformations once and reuse the same code for execution across multiple backends (e.g. pandas, Postgres, etc).
```
Comment (lostmygithubaccount) [inline docs/tutorials/open-source/apache-flink/0_setup.qmd:5]: I don't think this (or the next paragraph) are necessary -- would just say that in this tutorial you'll learn how to setup and use Flink w/ Ibis

File: docs/tutorials/open-source/apache-flink/0_setup.qmd [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+In today’s data infrastructures, it’s common for data scientists to do development locally and experiment with a sample dataset in Python. If the results look promising, the features and models would then be deployed into production. Oftentimes, there is a code handover in this process and a dedicated team of developers would be responsible for rewriting the logic for production. This incurs a huge amount of overhead.
+
+With the launch of support for the Flink backend in Ibis, we can now reuse the same code to work directly with streaming data in production. You can connect to data sources and sinks and write transformations in Python for both local experimentation and production deployment. Ibis provides a dataframe API that data scientists should already be familiar with, so that you don’t have to spend time learning a new query/processing language.
```
Comment (lostmygithubaccount) [inline docs/tutorials/open-source/apache-flink/0_setup.qmd:9]: also would rephrase more simply -- e.g. in the Starburst Galaxy tutorial we just have: "In this tutorial, we will connect to Starburst Galaxy and verify our connection. Following tutorials will go through the basics of Ibis on Starburst Galaxy’s demo data."

File: docs/tutorials/open-source/apache-flink/0_setup.qmd [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+
+## Getting started
+
+Follow [the instructions on the Ibis project page](https://ibis-project.org/install) to install the Flink backend for Ibis.
```
Comment (lostmygithubaccount) [inline docs/tutorials/open-source/apache-flink/0_setup.qmd:14]: this should link out to the proper tab or perhaps just duplicate install info for Flink here, otherwise could be jarring for a user to end up on the Install page w/ a lot of options

File: docs/tutorials/open-source/apache-flink/0_setup.qmd [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+```
+
+::: {.callout-tip}
+If you’re working on a batch data pipeline, simply change the TableEnvironment settings to batch mode before connecting to it:
+```{python}
+env_settings = EnvironmentSettings.in_batch_mode()
+```
+:::
+
+Now you can connect to data sources, create transformations, and write the results into sinks!
```
Comment (lostmygithubaccount) [inline docs/tutorials/open-source/apache-flink/0_setup.qmd:35]: link out to the next page in the tutorial

File: docs/_quarto.yml [hunk 8]
```diff
@@ -101,6 +101,7 @@ website:
         - install.qmd
         - auto: tutorials/*.qmd
         - auto: tutorials/data-platforms
```
Comment (lostmygithubaccount) [inline docs/_quarto.yml:103]: let's rename this to `cloud-data-platforms`

File: docs/tutorials/open-source/apache-flink/1_single_feature.qmd [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+
+
+### Prerequisites
+
+* Docker Compose: This tutorial uses Docker Compose to manage an Apache Kafka environment (including sample data generation) and a Flink cluster (for remote execution). You can [download and install Docker Compose from the official website](https://docs.docker.com/compose/install/).
+* JDK 11 release: Flink requires Java 11.
+* Python 3.9 or 3.10.
+* Follow [the instructions on the Ibis project page](https://ibis-project.org/install) to install the Flink backend for Ibis. For the tutorial below, we assume that you already have the Ibis package correctly installed in your environment.
```
Comment (lostmygithubaccount) [inline docs/tutorials/open-source/apache-flink/1_single_feature.qmd:11]: same feedback -- I would just link back to the setup part of the tutorial and assume the user has followed that for this page

**Review by lostmygithubaccount — COMMENTED** [review]
in general, wrap lines. I would keep the tutorial document a bit more simple and focused on getting the user up and running

File: docs/tutorials/open-source/apache-flink/1_single_feature.qmd [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+for _, msg in zip(range(10), consumer):
+    # this ensures that messages exist in the `transaction` topic before proceeding
+    print(msg)
+```
+
+::: {.callout-tip}
+Running `docker compose up` with the `-d` flag runs it in detached mode, where containers are run in the background. While this frees up the terminal for you to run other commands, it also hides error messages.
+
+Depending on whether the container images are already available locally, setting up the containers may take anywhere from 10 seconds to a minute. If it's your first time running this command, it's best to run it in the foreground so that you can monitor the progress of setup.
+
```
[the remark this answers is not in the document]
Comment (chloeh13q) [reply docs/tutorials/open-source/apache-flink/1_single_feature.qmd:38]: For some reason when I remove this line the callout-tip box doesn't get rendered appropriately in the preview

[the remark this answers is not in the document]
Comment (chloeh13q) [reply docs/tutorials/open-source/apache-flink/1_single_feature.qmd]: Nvm, it looks okay now!

[the remark this answers is not in the document]
Comment (chloeh13q) [reply docs/tutorials/open-source/apache-flink/1_single_feature.qmd]: Nvm, it looks okay now!

**Comment by chloeh13q** [comment]
@lostmygithubaccount Thanks for the review, I addressed all of your comments!

> I would keep the tutorial document a bit more simple and focused on getting the user up and running
The intention is to get to the point where the user can write a transformation. If this is too long, I can split it into two and have the first tutorial go up to the point where we write a simple over aggregation, and the second post cover window aggregation and what the differences are (but the setup code would be identical, i.e. connect to a source, connect to a sink). Alternatively, we can provide the code for over aggregation in the example repo but not explicitly talk about the syntax in this tutorial. What do you think?

P.S. Still pending porting the complete code example to the new ibis-flink-example repo.

**Comment by lostmygithubaccount** [comment]
will also need to create a redirect for the `data-platforms` -> `cloud-data-platforms` URL breakage

File: docs/_quarto.yml [hunk 11]
```diff
@@ -100,7 +100,8 @@ website:
       contents:
         - install.qmd
         - auto: tutorials/*.qmd
-        - auto: tutorials/data-platforms
+        - auto: tutorials/cloud-data-platforms
+        - auto: tutorials/open-source-frameworks
```
Comment (lostmygithubaccount) [inline docs/_quarto.yml:104]: apologies to request another change here -- can you switch this to "open-source-software" and add the redirect for the data platforms?

File: docs/tutorials/open-source-frameworks/apache-flink/0_setup.qmd [hunk 12]
```diff
@@ -0,0 +1,53 @@
+# Getting started
+
+In this tutorial, you will learn how to set up and use Flink with Ibis. If you
+want to see Ibis in action with a real-life example, check out [A real-life
+use case: fraud detection](1_single_feature.qmd).
```
Comment (lostmygithubaccount) [inline docs/tutorials/open-source-frameworks/apache-flink/0_setup.qmd:5]: I would slightly rephase the second sentence to something like "after setup in this part, we'll see a real-life example..." or something just so the user doesn't jump ahead before getting everything installed and whatnot

**Review by lostmygithubaccount — CHANGES_REQUESTED** [review]
one more request then I think this is good to merge

**Comment by chloeh13q** [comment]
@lostmygithubaccount BTW, we're also working on adding a visualization dashboard to the demo. We think it adds some fun elements and could help make the demo more interactive and tangible. We will provide the complete code for the dashboard so that all that the user needs to do is to spin up the service for the dashboard (via one line command) and then they should be able to see plots on a local port. Does that sound reasonable to you? (We can put this in a separate PR because we're still tweaking some parts)

**Comment by lostmygithubaccount** [comment]
sounds great! follow up PR seems fine for that

what kind of dashboard is this? Quarto?

**Comment by chloeh13q** [comment]
No it's a [Dash](https://dash.plotly.com/) app. It's not a part of the Quarto rendering - in the tutorial it would tell users to go to http://127.0.0.1:8050/ in the browser to see the dashboard. We're also thinking about adding a grafana dashboard for more complex visualizations, but won't tackle that now.

**Comment by lostmygithubaccount** [comment]
why dash? if it's quarto we can put it right in the website (like the [backend support matrix](https://ibis-project.org/support_matrix))

**Comment by chloeh13q** [comment]
From @mfatihaktas :
> For the following reasons:
- As much as I could understand it, Quarto is for reproducible technical reporting. It integrates very well with the notebooks and allows for quick implementation of beautiful graphs. However, I could not find much information on how to use it as a "web service" in a containerized setting. The alternative would to implement a "wrapper python module" around Quarto API and insert calls to it in the demo notebook. I thought that this would "pollute/inflate" and might deviate/conflate the goal of the notebook: introducing Ibis/Flink.
- Quarto's functionality seems to be limited in terms of developing dynamic graphs. It allows for responding to user input (e.g., selecting a button on the graphs), however I could not find enough information on how to generate semi-runtime-plots with it.
- Dash requires very small amount of boilerplate code to turn data into a dashboard. Thus, we can easily replace it with any other tool that allows for serving dashboards over HTTP.
We should be able to switch to any other viz tool quickly as long as the dashboard tool works in a containerized setting. Otherwise, we would need to figure out how to bring the dashboard into the demo notebook.

**Comment by mfatihaktas** [comment]
> From @mfatihaktas :
> 
> > For the following reasons:
> > 
> > * As much as I could understand it, Quarto is for reproducible technical reporting. It integrates very well with the notebooks and allows for quick implementation of beautiful graphs. However, I could not find much information on how to use it as a "web service" in a containerized setting. The alternative would to implement a "wrapper python module" around Quarto API and insert calls to it in the demo notebook. I thought that this would "pollute/inflate" and might deviate/conflate the goal of the notebook: introducing Ibis/Flink.
> > * Quarto's functionality seems to be limited in terms of developing dynamic graphs. It allows for responding to user input (e.g., selecting a button on the graphs), however I could not find enough information on how to generate semi-runtime-plots with it.
> > * Dash requires very small amount of boilerplate code to turn data into a dashboard. Thus, we can easily replace it with any other tool that allows for serving dashboards over HTTP.
> 
> > We should be able to switch to any other viz tool quickly as long as the dashboard tool works in a containerized setting. Otherwise, we would need to figure out how to bring the dashboard into the demo notebook.

Quick note. Dash apps can be [embedded in Quarto easily](https://community.plotly.com/t/embedding-dash-app-in-quarto/75864) as `dash` is powered by `plotly`.

**Review by lostmygithubaccount — APPROVED** [review]
thanks!
